### PR TITLE
Dataloader's cache: support invalidation and importing

### PR DESCRIPTION
**Pull request by paulo-raca**
A big part of the power of Dataloaders comes from its internal cache.

This PR aims to take improve it with 2 new features:
- Data invalidation: After a mutations, it is important to remove old data from the cache so we can continue with up-to-date data.
- Data insertion: We can now insert data retrieved externally (Usually by a direct database call) into the cache, avoiding duplicate fetching between places that use the database directly and places that use the dataloader.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This release improves the dataloader class with new features:

- Explicitly cache invalidation, prevents old data from being fetched after a mutation
- Importing data into the cache, prevents unnecessary load calls if the data has already been fetched by other means.


---
Here's the preview release card for twitter:




Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Paulo Costa for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: docs/guides/dataloaders.md [hunk 1]
```diff
@@ -118,6 +118,62 @@ the list for each incorrect key. A call with `keys == [1, 3]` returns
...
+### Cache invalidation
+
+By default DataLoaders use an internal cache. It is great for performance, however it can cause problems when the data is modified
+(i.e., a mutation), as the cached data is no longer be valid! 😮
+
+To fix it, you can explicitly invalidate the data in the cache, using one of these ways:
+
+- Specifying a key with `loader.invalidate(id)`,
+- Specifying several keys with `loader.invalidate_many([id1, id2, id3, ...])`,
+- Invalidating the whole cache with `loader.invalidate_all()`
```
Comment (patrick91) [inline docs/guides/dataloaders.md:130]: should we add a note that this is the same a clear in dataloader js?

and same for preset being prime in dataloader js?

Comment (paulo-raca) [reply docs/guides/dataloaders.md:130]: Good point!

I really didn't think about keeping consistency between this and the JS api, but it sounds like a good idea

Comment (paulo-raca) [reply docs/guides/dataloaders.md]: I renamed the method to `clear_*` and `prime_*`

File: docs/guides/dataloaders.md [hunk 2]
```diff
@@ -118,6 +118,62 @@ the list for each incorrect key. A call with `keys == [1, 3]` returns
 directly. If the `load_users` function raises an exception, even `load`s with an
 otherwise valid key, like `await loader.load(1)`, will raise that exception.
 
+### Cache invalidation
```
Comment (patrick91) [inline docs/guides/dataloaders.md:121]: doesn't look like the docs are updated with the new names 😊

Comment (patrick91) [reply docs/guides/dataloaders.md:121]: ah, it is missing docs about priming, I will add them

**Review by patrick91 — APPROVED** [review]
Nice one! Thank you 😊

**Comment by paulo-raca** [comment]
Thanks for looking into it 👍

**Pull request merged**